client/src/unifyfs-fixed.c

```diff
@@ -264,74 +264,209 @@ static int unifyfs_chunk_read(
 
-    } else {
-        /*
-        cur_slice_start                     cur_slice_endnext_slice_start                   next_slice_end
-                         cur_idx_start                                      cur_idx_end
+    /* extend length of last index to include beginning portion
+     * of current index */
+    prev_idx->length += length;
 
-        */
-        index_set->idxes[index_set->count] = *cur_idx;
-        index_set->idxes[index_set->count].length =
-            cur_slice_end - cur_idx_start + 1;
+    /* adjust current index to subtract off those bytes */
+    next_idx->file_pos += length;
+    next_idx->log_pos  += length;
+    next_idx->length   -= length;
 
-        cur_mem_pos += index_set->idxes[index_set->count].length;
+    return UNIFYFS_SUCCESS;
+}
 
-        cur_slice_start = cur_slice_end + 1;
-        cur_slice_end = cur_slice_start + slice_range - 1;
-        index_set->count++;
+/* given an index, split it into multiple indices whose range is equal or
+ * smaller than slice_range size
+ * @param  cur_idx:     the index to split
+ * @param  slice_range: the slice size of the key-value store
+ * @return index_set:   the set of split indices
+ * @param  maxcount:     number of entries in output array */
+static int unifyfs_split_index(
+    unifyfs_index_t* cur_idx,   /* write index to split (offset and length) */
+    long slice_range,           /* number of bytes in each slice */
+    unifyfs_index_t* index_set, /* output array to store new indexes in */
+    off_t maxcount,             /* max number of items in output array */
+    off_t* used_count)          /* number of entries we added in split */
+{
+    /* first byte offset this write will write to */
+    long idx_start = cur_idx->file_pos;
 
-        while (1) {
-            if (cur_idx_end <= cur_slice_end) {
-                break;
-            }
+    /* last byte offset this write will write to */
+    long idx_end = cur_idx->file_pos + cur_idx->length - 1;
 
-            index_set->idxes[index_set->count].fid = cur_idx->fid;
-            index_set->idxes[index_set->count].file_pos = cur_slice_start;
-            index_set->idxes[index_set->count].length = slice_range;
-            index_set->idxes[index_set->count].mem_pos = cur_mem_pos;
-            cur_mem_pos += index_set->idxes[index_set->count].length;
+    /* starting byte offset of slice that first write offset falls in */
+    long slice_start = (idx_start / slice_range) * slice_range;
 
-            cur_slice_start = cur_slice_end + 1;
-            cur_slice_end = cur_slice_start + slice_range - 1;
-            index_set->count++;
+    /* last byte offset of slice that first write offset falls in */
+    long slice_end = slice_start + slice_range - 1;
 
+    /* get pointer to first output index structure */
+    unifyfs_index_t* set = index_set;
+
+    /* initialize count of output index entries */
+    off_t count = 0;
+
+    /* no room to write index values */
+    if (count >= maxcount) {
```

Comment:
How can this condition ever be true? Seems only when `maxcount == 0`, in which case we shouldn't call this function in the first place.